lib/nrf_modem_lib/trace_medium.c

```diff
@@ -0,0 +1,115 @@
+/*
+ * Copyright (c) 2022 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <stddef.h>
+#include <string.h>
+#include <zephyr.h>
+#include <nrf_modem.h>
+#include <modem/trace_medium.h>
+#include <logging/log.h>
+
+LOG_MODULE_REGISTER(trace_medium, CONFIG_NRF_MODEM_LIB_LOG_LEVEL);
+
+static struct trace_medium_entry *active_medium;
+static struct trace_medium_entry *new_medium;
+static bool switch_active_medium;
+
+static void select_first_ready_medium(void)
+{
+	struct trace_medium_entry *medium = NULL;
+	const char *medium_name = "none";
+
+	STRUCT_SECTION_FOREACH(trace_medium_entry, e) {
+		if (e->ready) {
+			medium = e;
+			break;
+		}
+	}
+
+	if (medium != NULL) {
+		medium_name = medium->name;
+	}
+	LOG_INF("Active trace medium: %s", medium->name);
+
+	new_medium = medium;
+	switch_active_medium = true;
+}
+
+void trace_medium_init(void)
+{
+	int err;
+
+	STRUCT_SECTION_FOREACH(trace_medium_entry, e) {
+		if (e->init == NULL || e->write == NULL) {
+			e->ready = false;
+			continue;
+		}
+
+		err = e->init();
+		if (e->ready && (err != -EALREADY && err != 0)) {
+			e->ready = false;
+			LOG_ERR("Failed to reinit trace medium %s, err %d", e->name, err);
+		} else if (!e->ready && err) {
+			LOG_ERR("Failed to init trace medium %s, err %d", e->name, err);
+		} else {
+			e->ready = true;
+		}
+	}
+
+#if defined(CONFIG_NRF_MODEM_LIB_TRACE_MEDIUM_UART)
```

Comment:
Do you need this check? Couldn't the `select_first_ready_medium()` be used always?